groovy-language-server/src/main/java/com/palantir/ls/util/SourceWriter.java

```diff
@@ -0,0 +1,191 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.ls.util;
+
+import io.typefox.lsapi.Position;
+import io.typefox.lsapi.TextDocumentContentChangeEvent;
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.commons.io.FileUtils;
+
+/**
+ * Writes incremental changes to files.
+ */
+public final class SourceWriter {
+
+    private final Path source;
+    private final Path destination;
+
+    private SourceWriter(Path source, Path destination) {
+        this.source = source;
+        this.destination = destination;
+    }
+
+    /**
+     * Creates a SourceWriter which creates a copy of source at destination.
+     */
+    public static SourceWriter of(Path source, Path destination) throws IOException {
+        SourceWriter writer = new SourceWriter(source, destination);
+        writer.initialize();
+        return writer;
+    }
+
+    public Path getSource() {
+        return source;
+    }
+
+    public Path getDestination() {
+        return destination;
+    }
+
+    /**
+     * Applies the given changes to the destination file. Does not handle intersecting ranges in the changes.
+     */
+    public synchronized void applyChanges(List<TextDocumentContentChangeEvent> contentChanges) throws IOException {
+        // From earliest start of range to latest
+        List<TextDocumentContentChangeEvent> sortedChanges =
+                contentChanges.stream().sorted((c1, c2) -> Ranges.POSITION_COMPARATOR.compare(c1.getRange().getStart(),
+                        c2.getRange().getStart())).collect(Collectors.toList());
+
+        // TODO(#60): Check if any of the changes intersect and throw an exception
+        handleChanges(sortedChanges);
+    }
+
+    private synchronized void handleChanges(List<TextDocumentContentChangeEvent> sortedChanges)
+            throws IOException {
+        BufferedReader file =
+                new BufferedReader(new InputStreamReader(new FileInputStream(destination.toAbsolutePath().toString()),
+                        StandardCharsets.UTF_8.toString()));
+        File tempFile = File.createTempFile("tempdestination", ".tmp");
```

Comment:
`File.createTempFile` creates a random file name with the given prefix and suffix. For example in this case it would create something like: `tempdestination2868287298265030524.tmp`